schema/crdb/dbinit.sql

```diff
@@ -3545,14 +3545,29 @@ SELECT
   deleted
 FROM interleaved_versions;
 
+-- Table for optionally enabling / disbling background tasks or
+-- subsections of background tasks at runtime
+CREATE TABLE IF NOT EXISTS omicron.public.background_task_toggles (
+    id UUID NOT NULL DEFAULT uuid_generate_v4(),
+    name STRING NOT NULL,
+    enabled BOOLEAN NOT NULL,
+    time_created TIMESTAMPTZ NOT NULL DEFAULT NOW(),
+    time_modified TIMESTAMPTZ ON UPDATE NOW() NOT NULL DEFAULT NOW(),
+    time_deleted TIMESTAMPTZ
+);
+
+CREATE UNIQUE INDEX IF NOT EXISTS background_task_toggles_name
```

Comment:
> Why bother with an id here? Why not make name the primary key?
For some reason I recall there being a bit of discussion debating whether we should always use surrogate keys or if it was okay to opt for using natural keys. I don't really mind either way.

> (I think we generally have avoided default values in the database and specify them from Rust instead. Looking at dbinit.sql I see that we've been a little inconsistent about this but there are only 5 instances in that file where time_created or time_modified has DEFAULT NOW() compared with about 100 where they don't.)

I'm okay with changing this as well, but I guess I'm a bit fuzzy on how loose or strict we're being with our timestamps, as I see calls to `chrono::Utc::now()` as well as diesel's `dsl::some_timestamp_column.eq(Utc::now())` on the Rust side. I'm guessing when you say "set it in Rust" you're referring to doing it via a diesel query, which I'm guessing does `NOW()` in the query builder?

I'm asking about this distinction because I'm concerned about the following:

When a particular instance of nexus creates / modifies a record and sets the timestamp via Rust code via `chrono::Utc::now()`, that record has not yet been created / modified from the perspective of all of the other nexus instances. As far as I can tell, the _real_ `time_*` is when the DB commits the transaction for all of the other Nexus instances to see. When we let the DB set `time_created` and `time_modified`, we are getting a more accurate representation of when these events are visible and actionable from the perspective of all of the other Nexus instances.

Consider the following circumstance:
1. Nexus A timestamps a record, in Rust, with Timestamp1
2. Nexus B timestamps a record with Timestamp2
3. They both insert at roughly the same time, but for whatever reason (latency, disk, etc.) the DB commits the record from Nexus B _first_.
4. When operators / developers look at the DB, we are led to believe that the record from Nexus A was created _before_ Nexus B, but the truth is the that the record for Nexus B was there first, and could have been affecting application behavior before Nexus A.

Being that our application embraces a lot of idempotent & eventually-consistent patterns, I don't _think_ this will get us into trouble, but I think it's worth considering this scenario because it could create some confusing moments down the road whenever we try to use the timestamp data.

Reply:
Yeah, timestamps cannot be used for correctness for the reasons you mentioned.  Even when they're generated by the database, it doesn't reflect the time at which the data became visible to other transactions.  (That's not even well-defined, since different CockroachDB nodes will have slightly different times.)  I think the database evaluates `NOW()` when parsing or processing the query but decidedly before the transaction is committed.

I tried this: in one shell, `cockroach demo` and set up a table:

```
dap@zathras ~ $ cockroach demo
#
# Welcome to the CockroachDB demo database!
#
# You are connected to a temporary, in-memory CockroachDB cluster of 1 node.
#
# This demo session will attempt to enable enterprise features
# by acquiring a temporary license from Cockroach Labs in the background.
# To disable this behavior, set the environment variable
# COCKROACH_SKIP_ENABLING_DIAGNOSTIC_REPORTING=true.
#
# Beginning initialization of the movr dataset, please wait...
#
# The cluster has been preloaded with the "movr" dataset
# (MovR is a fictional vehicle sharing company).
#
# Reminder: your changes to data stored in the demo session will not be saved!
#
# If you wish to access this demo cluster using another tool, you will need
# the following details:
#
#   - Connection parameters:
#     (webui)    http://127.0.0.1:8080/demologin?password=demo75304&username=demo
#     (sql)      postgres://demo:demo75304@127.0.0.1:26257?sslmode=require
#     (sql/unix) postgres://demo:demo75304@?host=%2Fvar%2Ffolders%2Fx3%2F6tbrhnss6zz677z4zjsyvkqc0000gn%2FT%2Fdemo351111831&port=26257
#   
#   - Username: "demo", password: "demo75304"
#   - Directory with certificate files (for certain SQL drivers/tools): /var/folders/x3/6tbrhnss6zz677z4zjsyvkqc0000gn/T/demo351111831
#
# Server version: CockroachDB CCL v21.1.10 (x86_64-apple-darwin19, built 2021/10/07 02:42:10, go1.15.14) (same version as client)
# Cluster ID: 9034e646-3342-4a6a-b2cd-d0da89bc3ebe
# Organization: Cockroach Demo
#
# Enter \? for a brief introduction.
#
demo@127.0.0.1:26257/movr> create table t1 (c INT, v TIMESTAMPTZ);
CREATE TABLE

Time: 7ms total (execution 7ms / network 0ms)

```

In that shell, open a transaction, do an INSERT with `NOW()`, but don't commit it yet:

```
demo@127.0.0.1:26257/movr> BEGIN;
BEGIN

Time: 0ms total (execution 0ms / network 0ms)

demo@127.0.0.1:26257/movr  OPEN> INSERT INTO t1 (c, v) VALUES (1, NOW());
INSERT 1

Time: 1ms total (execution 1ms / network 0ms)


```

Now wait like 10 seconds.  Then in another shell, run `cockroach sql` to connect to the same database and do a similar INSERT but right away:

```
dap@zathras ~ $ cockroach sql --url 'postgres://demo:demo75304@127.0.0.1:26257?sslmode=require'
#
# Welcome to the CockroachDB SQL shell.
# All statements must be terminated by a semicolon.
# To exit, type: \q.
#
# Server version: CockroachDB CCL v21.1.10 (x86_64-apple-darwin19, built 2021/10/07 02:42:10, go1.15.14) (same version as client)
# Cluster ID: 9034e646-3342-4a6a-b2cd-d0da89bc3ebe
# Organization: Cockroach Demo
#
# Enter \? for a brief introduction.
#
demo@127.0.0.1:26257/defaultdb> use movr;
SET

Time: 1ms total (execution 1ms / network 0ms)

demo@127.0.0.1:26257/movr> INSERT INTO t1 (c, v) VALUES (2, NOW());
INSERT 1

Time: 2ms total (execution 2ms / network 0ms)

```

As soon as possible, in the first one, COMMIT:

```
demo@127.0.0.1:26257/movr  OPEN> COMMIT;
COMMIT

Time: 1ms total (execution 1ms / network 0ms)
```

So we did the INSERTs about 10 seconds apart but the commits about the same time.  This is what we get:

```
demo@127.0.0.1:26257/movr> SELECT * from t1;
  c |               v
----+--------------------------------
  1 | 2024-03-06 20:11:35.018453+00
  2 | 2024-03-06 20:11:49.489064+00
(2 rows)

Time: 1ms total (execution 1ms / network 0ms)
```

i.e., the timestamps are 14 seconds apart, which means the timestamp of the first one was 14 seconds before it was comitted.

Reply:
I tried to go further and show an intermediate state where the *later* row is visible before the earlier row is.  This is hard because of CockroachDB's serializeable guarantee.  What happens instead is that if you try to list the table, it blocks until you commit the open transaction.  I'm fairly sure I've seen this sort of thing with PostgreSQL but it's not as easy for me to test that.

Reply:
:open_mouth: 

_the entire world is a lie_

But seriously, thanks for taking the time to provide that demo! Yet another interesting nuance of building distributed systems! :joy: